### ISSUE TITLE
Some specs will fail randomly

### ISSUE THREAD

**Issue by SpComb** [kontena/kontena#1902]
## Server
```
  1) /v1/services POST /:id/deploy changes state to deploy_pending

     Failure/Error: expect {

       expected result to have changed from false to true, but did not change

     # ./spec/api/v1/services_spec.rb:314:in `block (3 levels) in <top (required)>'
     # /home/travis/.rvm/gems/ruby-2.3.3@global/gems/bundler-1.14.4/lib/bundler/cli/exec.rb:74:in `load'
     # /home/travis/.rvm/gems/ruby-2.3.3@global/gems/bundler-1.14.4/lib/bundler/cli/exec.rb:74:in `kernel_load'
     # /home/travis/.rvm/gems/ruby-2.3.3@global/gems/bundler-1.14.4/lib/bundler/cli/exec.rb:27:in `run'
     # /home/travis/.rvm/gems/ruby-2.3.3@global/gems/bundler-1.14.4/lib/bundler/cli.rb:335:in `exec'
     # /home/travis/.rvm/gems/ruby-2.3.3@global/gems/bundler-1.14.4/lib/bundler/vendor/thor/lib/thor/command.rb:27:in `run'
     # /home/travis/.rvm/gems/ruby-2.3.3@global/gems/bundler-1.14.4/lib/bundler/vendor/thor/lib/thor/invocation.rb:126:in `invoke_command'
     # /home/travis/.rvm/gems/ruby-2.3.3@global/gems/bundler-1.14.4/lib/bundler/vendor/thor/lib/thor.rb:359:in `dispatch'
     # /home/travis/.rvm/gems/ruby-2.3.3@global/gems/bundler-1.14.4/lib/bundler/cli.rb:20:in `dispatch'
     # /home/travis/.rvm/gems/ruby-2.3.3@global/gems/bundler-1.14.4/lib/bundler/vendor/thor/lib/thor/base.rb:440:in `start'
     # /home/travis/.rvm/gems/ruby-2.3.3@global/gems/bundler-1.14.4/lib/bundler/cli.rb:11:in `start'
     # /home/travis/.rvm/gems/ruby-2.3.3@global/gems/bundler-1.14.4/exe/bundle:32:in `block in <top (required)>'
     # /home/travis/.rvm/gems/ruby-2.3.3@global/gems/bundler-1.14.4/lib/bundler/friendly_errors.rb:121:in `with_friendly_errors'
     # /home/travis/.rvm/gems/ruby-2.3.3@global/gems/bundler-1.14.4/exe/bundle:24:in `<top (required)>'
```

**Comment by kke** [comment]
Sometimes mongo or connection to it fails:

```
An error occurred in an after hook
    NoMethodError: undefined method `length' for nil:NilClass
    occurred at /home/travis/build/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/moped-2.0.7/lib/moped/database.rb:56:in `block in collection_names'
F
  An error occurred in an after hook
    Moped::Errors::QueryFailure: The operation: #<Moped::Protocol::Query
  @length=108
  @request_id=7644
  @response_to=0
  @op_code=2004
  @flags=[]
  @full_collection_name="kontena_test.system.namespaces"
  @skip=0
  @limit=0
  @selector={:name=>{"$not"=>/kontena_test\.system\.|\$/}}
  @fields=nil>
failed with error 11000: "exception: E11000 duplicate key error index: kontena_test.distributed_locks.$name_1 dup key: { : \"foo1\" }"
``` 
(something like this is replicated to almost every test)

**Comment by kke** [comment]
Agent:
```
1) Kontena::Workers::EventWorker#stream_events retries if stream finishes
     Failure/Error: expect(spy).to receive(:check).exactly(2).times
       (Double :spy).check(any args)
           expected: 2 times with any arguments
           received: 1 time with any arguments
     # ./spec/lib/kontena/workers/event_worker_spec.rb:108:in `block (3 levels) in <top (required)>'
     # /home/travis/.rvm/gems/ruby-2.3.3@global/gems/bundler-1.14.5/lib/bundler/cli/exec.rb:74:in `load'
     # /home/travis/.rvm/gems/ruby-2.3.3@global/gems/bundler-1.14.5/lib/bundler/cli/exec.rb:74:in `kernel_load'
     # /home/travis/.rvm/gems/ruby-2.3.3@global/gems/bundler-1.14.5/lib/bundler/cli/exec.rb:27:in `run'
     # /home/travis/.rvm/gems/ruby-2.3.3@global/gems/bundler-1.14.5/lib/bundler/cli.rb:335:in `exec'
     # /home/travis/.rvm/gems/ruby-2.3.3@global/gems/bundler-1.14.5/lib/bundler/vendor/thor/lib/thor/command.rb:27:in `run'
     # /home/travis/.rvm/gems/ruby-2.3.3@global/gems/bundler-1.14.5/lib/bundler/vendor/thor/lib/thor/invocation.rb:126:in `invoke_command'
     # /home/travis/.rvm/gems/ruby-2.3.3@global/gems/bundler-1.14.5/lib/bundler/vendor/thor/lib/thor.rb:359:in `dispatch'
     # /home/travis/.rvm/gems/ruby-2.3.3@global/gems/bundler-1.14.5/lib/bundler/cli.rb:20:in `dispatch'
     # /home/travis/.rvm/gems/ruby-2.3.3@global/gems/bundler-1.14.5/lib/bundler/vendor/thor/lib/thor/base.rb:440:in `start'
     # /home/travis/.rvm/gems/ruby-2.3.3@global/gems/bundler-1.14.5/lib/bundler/cli.rb:11:in `start'
     # /home/travis/.rvm/gems/ruby-2.3.3@global/gems/bundler-1.14.5/exe/bundle:32:in `block in <top (required)>'
     # /home/travis/.rvm/gems/ruby-2.3.3@global/gems/bundler-1.14.5/lib/bundler/friendly_errors.rb:121:in `with_friendly_errors'
     # /home/travis/.rvm/gems/ruby-2.3.3@global/gems/bundler-1.14.5/exe/bundle:24:in `<top (required)>'
```

**Comment by kke** [comment]
Server:

```
 1) GridServiceSchedulerWorker#check_deploy_queue picks oldest item from deploy queue
     Failure/Error: expect(newest.reload.started_at).to be_nil
       expected: nil
            got: Wed, 01 Mar 2017 13:26:48 +0000
     # ./spec/jobs/grid_service_scheduler_worker_spec.rb:17:in `block (3 levels) in <top (required)>'
     # /home/travis/.rvm/gems/ruby-2.3.3@global/gems/bundler-1.14.5/lib/bundler/cli/exec.rb:74:in `load'
     # /home/travis/.rvm/gems/ruby-2.3.3@global/gems/bundler-1.14.5/lib/bundler/cli/exec.rb:74:in `kernel_load'
     # /home/travis/.rvm/gems/ruby-2.3.3@global/gems/bundler-1.14.5/lib/bundler/cli/exec.rb:27:in `run'
     # /home/travis/.rvm/gems/ruby-2.3.3@global/gems/bundler-1.14.5/lib/bundler/cli.rb:335:in `exec'
     # /home/travis/.rvm/gems/ruby-2.3.3@global/gems/bundler-1.14.5/lib/bundler/vendor/thor/lib/thor/command.rb:27:in `run'
     # /home/travis/.rvm/gems/ruby-2.3.3@global/gems/bundler-1.14.5/lib/bundler/vendor/thor/lib/thor/invocation.rb:126:in `invoke_command'
     # /home/travis/.rvm/gems/ruby-2.3.3@global/gems/bundler-1.14.5/lib/bundler/vendor/thor/lib/thor.rb:359:in `dispatch'
     # /home/travis/.rvm/gems/ruby-2.3.3@global/gems/bundler-1.14.5/lib/bundler/cli.rb:20:in `dispatch'
     # /home/travis/.rvm/gems/ruby-2.3.3@global/gems/bundler-1.14.5/lib/bundler/vendor/thor/lib/thor/base.rb:440:in `start'
     # /home/travis/.rvm/gems/ruby-2.3.3@global/gems/bundler-1.14.5/lib/bundler/cli.rb:11:in `start'
     # /home/travis/.rvm/gems/ruby-2.3.3@global/gems/bundler-1.14.5/exe/bundle:32:in `block in <top (required)>'
     # /home/travis/.rvm/gems/ruby-2.3.3@global/gems/bundler-1.14.5/lib/bundler/friendly_errors.rb:121:in `with_friendly_errors'
     # /home/travis/.rvm/gems/ruby-2.3.3@global/gems/bundler-1.14.5/exe/bundle:24:in `<top (required)>'
```

**Comment by SpComb** [comment]
Since we're now gathering all the flaky specs in one issue:

```
  1) Kontena::Launchers::Etcd#pull_image pulls image if it does not exist
     Failure/Error: expect(Docker::Image).to receive(:create).with({'fromImage' => image})
       (<Docker::Image (class)>).create({"fromImage"=>"kontena/etcd:2.2.4"})
           expected: 1 time with arguments: ({"fromImage"=>"kontena/etcd:2.2.4"})
           received: 0 times

     # ./spec/lib/kontena/launchers/etcd_launcher_spec.rb:82:in `block (3 levels) in <top (required)>'
```

**Comment by SpComb** [comment]
## Server

```
  1) ServiceBalancerJob#pending_deploys? returns true if pending deploys

     Failure/Error: expect(subject.pending_deploys?(service)).to be_truthy
       expected: truthy value
            got: false

     # ./spec/jobs/service_balancer_job_spec.rb:189:in `block (3 levels) in <top (required)>'
```

**Comment by kke** [comment]
### Server
```
  1) DistributedLock.with_lock returns false if getting lock timeouts
     Failure/Error: described_class.with_lock('foo1') { sleep 0.1 }
     Moped::Errors::OperationFailure:
       The operation: #<Moped::Protocol::Command
         @length=153
         @request_id=3271
         @response_to=0
         @op_code=2004
         @flags=[]
         @full_collection_name="kontena_test.$cmd"
         @skip=0
         @limit=-1
         @selector={:count=>"distributed_locks", :query=>{"name"=>"foo1", "lock_id"=>"6f710ba36c9ed15694856f1c67416ed0"}}
         @fields=nil>
       failed with error 11000: "exception: E11000 duplicate key error index: kontena_test.distributed_locks.$name_1 dup key: { : \"foo1\" }"
       
       See https://github.com/mongodb/mongo/blob/master/docs/errors.md
       for details about this error.
     # ./vendor/bundle/ruby/2.3.0/gems/moped-2.0.7/lib/moped/operation/read.rb:63:in `block in execute'
     # ./vendor/bundle/ruby/2.3.0/gems/moped-2.0.7/lib/moped/node.rb:608:in `block (3 levels) in flush'
     # ./vendor/bundle/ruby/2.3.0/gems/moped-2.0.7/lib/moped/node.rb:607:in `map'
     # ./vendor/bundle/ruby/2.3.0/gems/moped-2.0.7/lib/moped/node.rb:607:in `block (2 levels) in flush'
     # ./vendor/bundle/ruby/2.3.0/gems/moped-2.0.7/lib/moped/node.rb:193:in `block in ensure_connected'
     # ./vendor/bundle/ruby/2.3.0/gems/moped-2.0.7/lib/moped/node.rb:118:in `block in connection'
     # ./vendor/bundle/ruby/2.3.0/gems/connection_pool-2.2.0/lib/connection_pool.rb:64:in `block (2 levels) in with'
     # ./vendor/bundle/ruby/2.3.0/gems/connection_pool-2.2.0/lib/connection_pool.rb:63:in `handle_interrupt'
     # ./vendor/bundle/ruby/2.3.0/gems/connection_pool-2.2.0/lib/connection_pool.rb:63:in `block in with'
     # ./vendor/bundle/ruby/2.3.0/gems/connection_pool-2.2.0/lib/connection_pool.rb:60:in `handle_interrupt'
     # ./vendor/bundle/ruby/2.3.0/gems/connection_pool-2.2.0/lib/connection_pool.rb:60:in `with'
     # ./vendor/bundle/ruby/2.3.0/gems/moped-2.0.7/lib/moped/node.rb:116:in `connection'
     # ./vendor/bundle/ruby/2.3.0/gems/moped-2.0.7/lib/moped/node.rb:189:in `ensure_connected'
     # ./vendor/bundle/ruby/2.3.0/gems/moped-2.0.7/lib/moped/node.rb:603:in `block in flush'
     # ./vendor/bundle/ruby/2.3.0/gems/moped-2.0.7/lib/moped/node.rb:632:in `block in logging'
     # ./vendor/bundle/ruby/2.3.0/gems/activesupport-4.1.14/lib/active_support/notifications.rb:159:in `block in instrument'
     # ./vendor/bundle/ruby/2.3.0/gems/activesupport-4.1.14/lib/active_support/notifications/instrumenter.rb:20:in `instrument'
     # ./vendor/bundle/ruby/2.3.0/gems/activesupport-4.1.14/lib/active_support/notifications.rb:159:in `instrument'
     # ./vendor/bundle/ruby/2.3.0/gems/moped-2.0.7/lib/moped/instrumentable.rb:31:in `instrument'
     # ./vendor/bundle/ruby/2.3.0/gems/moped-2.0.7/lib/moped/node.rb:631:in `logging'
     # ./vendor/bundle/ruby/2.3.0/gems/moped-2.0.7/lib/moped/node.rb:602:in `flush'
     # ./vendor/bundle/ruby/2.3.0/gems/moped-2.0.7/lib/moped/node.rb:406:in `process'
     # ./vendor/bundle/ruby/2.3.0/gems/moped-2.0.7/lib/moped/operation/read.rb:48:in `execute'
     # ./vendor/bundle/ruby/2.3.0/gems/moped-2.0.7/lib/moped/node.rb:663:in `read'
     # ./vendor/bundle/ruby/2.3.0/gems/moped-2.0.7/lib/moped/node.rb:90:in `command'
     # ./vendor/bundle/ruby/2.3.0/gems/moped-2.0.7/lib/moped/database.rb:73:in `block in command'
     # ./vendor/bundle/ruby/2.3.0/gems/moped-2.0.7/lib/moped/cluster.rb:249:in `block in with_primary'
     # ./vendor/bundle/ruby/2.3.0/gems/moped-2.0.7/lib/moped/node.rb:219:in `block in ensure_primary'
     # ./vendor/bundle/ruby/2.3.0/gems/moped-2.0.7/lib/moped/executable.rb:25:in `execute'
     # ./vendor/bundle/ruby/2.3.0/gems/moped-2.0.7/lib/moped/node.rb:218:in `ensure_primary'
     # ./vendor/bundle/ruby/2.3.0/gems/moped-2.0.7/lib/moped/cluster.rb:248:in `with_primary'
     # ./vendor/bundle/ruby/2.3.0/gems/moped-2.0.7/lib/moped/read_preference/primary.rb:55:in `block in with_node'
     # ./vendor/bundle/ruby/2.3.0/gems/moped-2.0.7/lib/moped/retryable.rb:30:in `with_retry'
     # ./vendor/bundle/ruby/2.3.0/gems/moped-2.0.7/lib/moped/read_preference/primary.rb:54:in `with_node'
     # ./vendor/bundle/ruby/2.3.0/gems/moped-2.0.7/lib/moped/database.rb:72:in `command'
     # ./vendor/bundle/ruby/2.3.0/gems/moped-2.0.7/lib/moped/query.rb:42:in `count'
     # ./vendor/bundle/ruby/2.3.0/gems/mongoid-4.0.2/lib/mongoid/contextual/mongo.rb:60:in `block in count'
     # ./vendor/bundle/ruby/2.3.0/gems/mongoid-4.0.2/lib/mongoid/contextual/mongo.rb:457:in `try_cache'
     # ./vendor/bundle/ruby/2.3.0/gems/mongoid-4.0.2/lib/mongoid/contextual/mongo.rb:60:in `count'
     # ./vendor/bundle/ruby/2.3.0/gems/mongoid-4.0.2/lib/mongoid/contextual/mongo.rb:87:in `destroy'
     # ./vendor/bundle/ruby/2.3.0/gems/mongoid-4.0.2/lib/mongoid/contextual.rb:20:in `destroy'
     # ./app/models/distributed_lock.rb:53:in `release_lock'
     # ./app/models/distributed_lock.rb:28:in `with_lock'
     # ./spec/models/distributed_lock_spec.rb:28:in `block (4 levels) in <top (required)>'
  2) DistributedLock.with_lock locks access between threads
     Failure/Error: expect(results[0]).to eq('first_lock')
       
       expected: "first_lock"
            got: nil
       
       (compared using ==)
     # ./spec/models/distributed_lock_spec.rb:21:in `block (4 levels) in <top (required)>'
     # ./spec/models/distributed_lock_spec.rb:13:in `times'
     # ./spec/models/distributed_lock_spec.rb:13:in `block (3 levels) in <top (required)>'
     # /home/travis/.rvm/gems/ruby-2.3.3@global/gems/bundler-1.14.5/lib/bundler/cli/exec.rb:74:in `load'
     # /home/travis/.rvm/gems/ruby-2.3.3@global/gems/bundler-1.14.5/lib/bundler/cli/exec.rb:74:in `kernel_load'
     # /home/travis/.rvm/gems/ruby-2.3.3@global/gems/bundler-1.14.5/lib/bundler/cli/exec.rb:27:in `run'
     # /home/travis/.rvm/gems/ruby-2.3.3@global/gems/bundler-1.14.5/lib/bundler/cli.rb:335:in `exec'
     # /home/travis/.rvm/gems/ruby-2.3.3@global/gems/bundler-1.14.5/lib/bundler/vendor/thor/lib/thor/command.rb:27:in `run'
     # /home/travis/.rvm/gems/ruby-2.3.3@global/gems/bundler-1.14.5/lib/bundler/vendor/thor/lib/thor/invocation.rb:126:in `invoke_command'
     # /home/travis/.rvm/gems/ruby-2.3.3@global/gems/bundler-1.14.5/lib/bundler/vendor/thor/lib/thor.rb:359:in `dispatch'
     # /home/travis/.rvm/gems/ruby-2.3.3@global/gems/bundler-1.14.5/lib/bundler/cli.rb:20:in `dispatch'
     # /home/travis/.rvm/gems/ruby-2.3.3@global/gems/bundler-1.14.5/lib/bundler/vendor/thor/lib/thor/base.rb:440:in `start'
     # /home/travis/.rvm/gems/ruby-2.3.3@global/gems/bundler-1.14.5/lib/bundler/cli.rb:11:in `start'
     # /home/travis/.rvm/gems/ruby-2.3.3@global/gems/bundler-1.14.5/exe/bundle:32:in `block in <top (required)>'
     # /home/travis/.rvm/gems/ruby-2.3.3@global/gems/bundler-1.14.5/lib/bundler/friendly_errors.rb:121:in `with_friendly_errors'
     # /home/travis/.rvm/gems/ruby-2.3.3@global/gems/bundler-1.14.5/exe/bundle:24:in `<top (required)>'
```

**Comment by SpComb** [comment]
## Server

```
  1) GridServiceDeployer#deploy subscribes to ping

     Failure/Error: sleep 0.05 until events.size > 0

     Timeout::Error:

       execution expired

     # ./spec/services/grid_service_deployer_spec.rb:102:in `sleep'
     # ./spec/services/grid_service_deployer_spec.rb:102:in `block (4 levels) in <top (required)>'
     # ./spec/services/grid_service_deployer_spec.rb:101:in `block (3 levels) in <top (required)>'
```

https://travis-ci.org/kontena/kontena/jobs/208291791#L669

**Comment by SpComb** [comment]
## Server

```
  1) AuthProvider should configure itself from the Configuration

     Failure/Error: Configuration.destroy_all

     NoMethodError:

       undefined method `[]' for nil:NilClass

     # ./vendor/bundle/ruby/2.3.0/gems/moped-2.0.7/lib/moped/protocol/reply.rb:71:in `command_failure?'

     # ./vendor/bundle/ruby/2.3.0/gems/moped-2.0.7/lib/moped/protocol/command.rb:26:in `failure?'

     # ./vendor/bundle/ruby/2.3.0/gems/moped-2.0.7/lib/moped/operation/read.rb:62:in `block in execute'

     # ./vendor/bundle/ruby/2.3.0/gems/moped-2.0.7/lib/moped/node.rb:608:in `block (3 levels) in flush'

     # ./vendor/bundle/ruby/2.3.0/gems/moped-2.0.7/lib/moped/node.rb:607:in `map'

     # ./vendor/bundle/ruby/2.3.0/gems/moped-2.0.7/lib/moped/node.rb:607:in `block (2 levels) in flush'

     # ./vendor/bundle/ruby/2.3.0/gems/moped-2.0.7/lib/moped/node.rb:193:in `block in ensure_connected'

     # ./vendor/bundle/ruby/2.3.0/gems/moped-2.0.7/lib/moped/node.rb:118:in `block in connection'

     # ./vendor/bundle/ruby/2.3.0/gems/connection_pool-2.2.0/lib/connection_pool.rb:64:in `block (2 levels) in with'

     # ./vendor/bundle/ruby/2.3.0/gems/connection_pool-2.2.0/lib/connection_pool.rb:63:in `handle_interrupt'

     # ./vendor/bundle/ruby/2.3.0/gems/connection_pool-2.2.0/lib/connection_pool.rb:63:in `block in with'

     # ./vendor/bundle/ruby/2.3.0/gems/connection_pool-2.2.0/lib/connection_pool.rb:60:in `handle_interrupt'

     # ./vendor/bundle/ruby/2.3.0/gems/connection_pool-2.2.0/lib/connection_pool.rb:60:in `with'

     # ./vendor/bundle/ruby/2.3.0/gems/moped-2.0.7/lib/moped/node.rb:116:in `connection'

     # ./vendor/bundle/ruby/2.3.0/gems/moped-2.0.7/lib/moped/node.rb:189:in `ensure_connected'

     # ./vendor/bundle/ruby/2.3.0/gems/moped-2.0.7/lib/moped/node.rb:603:in `block in flush'

     # ./vendor/bundle/ruby/2.3.0/gems/moped-2.0.7/lib/moped/node.rb:632:in `block in logging'

     # ./vendor/bundle/ruby/2.3.0/gems/activesupport-4.1.14/lib/active_support/notifications.rb:159:in `block in instrument'

     # ./vendor/bundle/ruby/2.3.0/gems/activesupport-4.1.14/lib/active_support/notifications/instrumenter.rb:20:in `instrument'

     # ./vendor/bundle/ruby/2.3.0/gems/activesupport-4.1.14/lib/active_support/notifications.rb:159:in `instrument'

     # ./vendor/bundle/ruby/2.3.0/gems/moped-2.0.7/lib/moped/instrumentable.rb:31:in `instrument'

     # ./vendor/bundle/ruby/2.3.0/gems/moped-2.0.7/lib/moped/node.rb:631:in `logging'

     # ./vendor/bundle/ruby/2.3.0/gems/moped-2.0.7/lib/moped/node.rb:602:in `flush'

     # ./vendor/bundle/ruby/2.3.0/gems/moped-2.0.7/lib/moped/node.rb:373:in `pipeline'

     # ./vendor/bundle/ruby/2.3.0/gems/moped-2.0.7/lib/moped/operation/write.rb:47:in `execute'

     # ./vendor/bundle/ruby/2.3.0/gems/moped-2.0.7/lib/moped/node.rb:680:in `write'

     # ./vendor/bundle/ruby/2.3.0/gems/moped-2.0.7/lib/moped/node.rb:475:in `remove'

     # ./vendor/bundle/ruby/2.3.0/gems/moped-2.0.7/lib/moped/query.rb:327:in `block (2 levels) in remove'

     # ./vendor/bundle/ruby/2.3.0/gems/moped-2.0.7/lib/moped/cluster.rb:249:in `block in with_primary'

     # ./vendor/bundle/ruby/2.3.0/gems/moped-2.0.7/lib/moped/node.rb:219:in `block in ensure_primary'

     # ./vendor/bundle/ruby/2.3.0/gems/moped-2.0.7/lib/moped/executable.rb:25:in `execute'

     # ./vendor/bundle/ruby/2.3.0/gems/moped-2.0.7/lib/moped/node.rb:218:in `ensure_primary'

     # ./vendor/bundle/ruby/2.3.0/gems/moped-2.0.7/lib/moped/cluster.rb:248:in `with_primary'

     # ./vendor/bundle/ruby/2.3.0/gems/moped-2.0.7/lib/moped/query.rb:326:in `block in remove'

     # ./vendor/bundle/ruby/2.3.0/gems/moped-2.0.7/lib/moped/retryable.rb:30:in `with_retry'

     # ./vendor/bundle/ruby/2.3.0/gems/moped-2.0.7/lib/moped/query.rb:325:in `remove'

     # ./vendor/bundle/ruby/2.3.0/gems/mongoid-4.0.2/lib/mongoid/query_cache.rb:117:in `remove_with_clear_cache'

     # ./vendor/bundle/ruby/2.3.0/gems/mongoid-4.0.2/lib/mongoid/persistable/deletable.rb:81:in `delete_as_root'

     # ./vendor/bundle/ruby/2.3.0/gems/mongoid-4.0.2/lib/mongoid/persistable/deletable.rb:27:in `block in delete'

     # ./vendor/bundle/ruby/2.3.0/gems/mongoid-4.0.2/lib/mongoid/persistable/deletable.rb:115:in `prepare_delete'

     # ./vendor/bundle/ruby/2.3.0/gems/mongoid-4.0.2/lib/mongoid/persistable/deletable.rb:23:in `delete'

     # ./vendor/bundle/ruby/2.3.0/gems/mongoid-4.0.2/lib/mongoid/persistable/destroyable.rb:24:in `block in destroy'

     # ./vendor/bundle/ruby/2.3.0/gems/activesupport-4.1.14/lib/active_support/callbacks.rb:82:in `run_callbacks'

     # ./vendor/bundle/ruby/2.3.0/gems/mongoid-4.0.2/lib/mongoid/interceptable.rb:138:in `run_callbacks'

     # ./vendor/bundle/ruby/2.3.0/gems/mongoid-4.0.2/lib/mongoid/persistable/destroyable.rb:24:in `destroy'

     # ./vendor/bundle/ruby/2.3.0/gems/mongoid-4.0.2/lib/mongoid/persistable/destroyable.rb:54:in `block in destroy_all'

     # ./vendor/bundle/ruby/2.3.0/gems/mongoid-4.0.2/lib/mongoid/contextual/mongo.rb:658:in `yield_document'

     # ./vendor/bundle/ruby/2.3.0/gems/mongoid-4.0.2/lib/mongoid/contextual/mongo.rb:123:in `block in each'

     # ./vendor/bundle/ruby/2.3.0/gems/moped-2.0.7/lib/moped/cursor.rb:29:in `block in each'

     # ./vendor/bundle/ruby/2.3.0/gems/moped-2.0.7/lib/moped/cursor.rb:29:in `each'

     # ./vendor/bundle/ruby/2.3.0/gems/moped-2.0.7/lib/moped/cursor.rb:29:in `each'

     # ./vendor/bundle/ruby/2.3.0/gems/moped-2.0.7/lib/moped/query.rb:78:in `each'

     # ./vendor/bundle/ruby/2.3.0/gems/mongoid-4.0.2/lib/mongoid/contextual/mongo.rb:122:in `each'

     # ./vendor/bundle/ruby/2.3.0/gems/mongoid-4.0.2/lib/mongoid/contextual.rb:20:in `each'

     # ./vendor/bundle/ruby/2.3.0/gems/mongoid-4.0.2/lib/mongoid/persistable/destroyable.rb:54:in `destroy_all'

     # ./spec/services/auth_provider_spec.rb:7:in `block (2 levels) in <top (required)>'

     # /home/travis/.rvm/gems/ruby-2.3.3@global/gems/bundler-1.14.6/lib/bundler/cli/exec.rb:74:in `load'

     # /home/travis/.rvm/gems/ruby-2.3.3@global/gems/bundler-1.14.6/lib/bundler/cli/exec.rb:74:in `kernel_load'

     # /home/travis/.rvm/gems/ruby-2.3.3@global/gems/bundler-1.14.6/lib/bundler/cli/exec.rb:27:in `run'

     # /home/travis/.rvm/gems/ruby-2.3.3@global/gems/bundler-1.14.6/lib/bundler/cli.rb:335:in `exec'

     # /home/travis/.rvm/gems/ruby-2.3.3@global/gems/bundler-1.14.6/lib/bundler/vendor/thor/lib/thor/command.rb:27:in `run'

     # /home/travis/.rvm/gems/ruby-2.3.3@global/gems/bundler-1.14.6/lib/bundler/vendor/thor/lib/thor/invocation.rb:126:in `invoke_command'

     # /home/travis/.rvm/gems/ruby-2.3.3@global/gems/bundler-1.14.6/lib/bundler/vendor/thor/lib/thor.rb:359:in `dispatch'

     # /home/travis/.rvm/gems/ruby-2.3.3@global/gems/bundler-1.14.6/lib/bundler/cli.rb:20:in `dispatch'

     # /home/travis/.rvm/gems/ruby-2.3.3@global/gems/bundler-1.14.6/lib/bundler/vendor/thor/lib/thor/base.rb:440:in `start'

     # /home/travis/.rvm/gems/ruby-2.3.3@global/gems/bundler-1.14.6/lib/bundler/cli.rb:11:in `start'

     # /home/travis/.rvm/gems/ruby-2.3.3@global/gems/bundler-1.14.6/exe/bundle:32:in `block in <top (required)>'

     # /home/travis/.rvm/gems/ruby-2.3.3@global/gems/bundler-1.14.6/lib/bundler/friendly_errors.rb:121:in `with_friendly_errors'

     # /home/travis/.rvm/gems/ruby-2.3.3@global/gems/bundler-1.14.6/exe/bundle:24:in `<top (required)>'
```

**Comment by SpComb** [comment]
Server mongodb rspec errors might be caused by #1964

**Comment by kke** [comment]
### Server:

```
  1) /v1/users POST /:username/roles adds user to role
     Failure/Error: user = User.create!(email: 'john@domain.com', id: '123456')
     KeyError:
       key not found: "a"
```

**Comment by SpComb** [comment]
## Server
```
  1) Stacks::Deploy#run removes services that are removed from a stack

     Failure/Error: expect(GridService.find(redis.id)).to be_nil

       expected: nil

            got: #<GridService _id: 58d137a00b77f51d96000d64, created_at: 2017-03-21 14:24:32 UTC, updated_at: 2017-03-21 14:24:32 UTC, image_name: "redis:2.8", labels: {}, affinity: [], name: "redis", stateful: true, user: nil, container_count: 1, cmd: nil, entrypoint: nil, ports: [], env: [], memory: nil, memory_swap: nil, cpu_shares: nil, volumes: [], volumes_from: [], privileged: nil, cap_add: [], cap_drop: [], net: "bridge", state: "initialized", log_driver: nil, log_opts: {}, devices: [], pid: nil, deploy_requested_at: nil, deployed_at: nil, revision: 1, stack_revision: nil, strategy: "ha", grid_id: BSON::ObjectId('58d137a00b77f51d96000d5f'), image_id: nil, stack_id: BSON::ObjectId('58d137a00b77f51d96000d62'), network_ids: [BSON::ObjectId('58d137a00b77f51d96000d60')]>

     # ./spec/mutations/stacks/deploy_spec.rb:76:in `block (3 levels) in <top (required)>'
```

**Comment by kke** [comment]
We should grab a couple of random seeds from these failed runs and see if it always fails when running with the same seed.

**Comment by jnummelin** [comment]
```
Failures:
  1) Kontena::Workers::NodeInfoWorker#calculate_container_time container is running calculates container time since container is started
     Failure/Error: expect(time).to eq(50)
       
       expected: 50
            got: 51
       
       (compared using ==)
     # ./spec/lib/kontena/workers/node_info_worker_spec.rb:156:in `block (4 levels) in <top (required)>'
     # /home/travis/.rvm/gems/ruby-2.3.3@global/gems/bundler-1.14.6/lib/bundler/cli/exec.rb:74:in `load'
     # /home/travis/.rvm/gems/ruby-2.3.3@global/gems/bundler-1.14.6/lib/bundler/cli/exec.rb:74:in `kernel_load'
     # /home/travis/.rvm/gems/ruby-2.3.3@global/gems/bundler-1.14.6/lib/bundler/cli/exec.rb:27:in `run'
     # /home/travis/.rvm/gems/ruby-2.3.3@global/gems/bundler-1.14.6/lib/bundler/cli.rb:335:in `exec'
     # /home/travis/.rvm/gems/ruby-2.3.3@global/gems/bundler-1.14.6/lib/bundler/vendor/thor/lib/thor/command.rb:27:in `run'
     # /home/travis/.rvm/gems/ruby-2.3.3@global/gems/bundler-1.14.6/lib/bundler/vendor/thor/lib/thor/invocation.rb:126:in `invoke_command'
     # /home/travis/.rvm/gems/ruby-2.3.3@global/gems/bundler-1.14.6/lib/bundler/vendor/thor/lib/thor.rb:359:in `dispatch'
     # /home/travis/.rvm/gems/ruby-2.3.3@global/gems/bundler-1.14.6/lib/bundler/cli.rb:20:in `dispatch'
     # /home/travis/.rvm/gems/ruby-2.3.3@global/gems/bundler-1.14.6/lib/bundler/vendor/thor/lib/thor/base.rb:440:in `start'
     # /home/travis/.rvm/gems/ruby-2.3.3@global/gems/bundler-1.14.6/lib/bundler/cli.rb:11:in `start'
     # /home/travis/.rvm/gems/ruby-2.3.3@global/gems/bundler-1.14.6/exe/bundle:32:in `block in <top (required)>'
     # /home/travis/.rvm/gems/ruby-2.3.3@global/gems/bundler-1.14.6/lib/bundler/friendly_errors.rb:121:in `with_friendly_errors'
     # /home/travis/.rvm/gems/ruby-2.3.3@global/gems/bundler-1.14.6/exe/bundle:24:in `<top (required)>'
Finished in 10.01 seconds (files took 0.60641 seconds to load)
404 examples, 1 failure
Failed examples:
rspec ./spec/lib/kontena/workers/node_info_worker_spec.rb:149 # Kontena::Workers::NodeInfoWorker#calculate_container_time container is running calculates container time since container is started
Randomized with seed 60580
```

**Comment by SpComb** [comment]
## Server
https://travis-ci.org/kontena/kontena/jobs/219646661#L731
```
  1) DistributedLock.with_lock returns false if getting lock timeouts

     Failure/Error: expect(described_class.with_lock('foo1', 0.01) { "thread B got lock" }).to eq(false)

       

       expected: false

            got: "thread B got lock"

       

       (compared using ==)

     # ./spec/models/distributed_lock_spec.rb:31:in `block (4 levels) in <top (required)>'
```

**Comment by SpComb** [comment]
## Server

```
  1) /v1/services/:id/container_logs GET when since parameter is passed returns service container logs created after passed timestamp

     Failure/Error: expect(json_response['logs'].size).to eq(2)

       

       expected: 2

            got: 1

       

       (compared using ==)

     # ./spec/api/v1/service_container_logs_spec.rb:83:in `block (4 levels) in <top (required)>'
```

**Comment by SpComb** [comment]
## Server
```
  1) GridServiceDeployer for a grid with host nodes #deploy for multiple concurrent instance deploys fails the service deploy if one of the concurrent instance deploys fail

     Failure/Error: expect{

       expected result to have changed to 2, but is now 1

     # ./spec/services/grid_service_deployer_spec.rb:162:in `block (5 levels) in <top (required)>'
```

**Comment by SpComb** [comment]
Most of the server rspec failures in this issue look like they will have been fixed in #1987 and #2284.

Closing this, open up new issues for each individual spec that continues to be flaky. There's some in this issue that look like they're probably still valid.
